commcare_connect/users/user_credentials.py

```diff
@@ -0,0 +1,108 @@
+            level=credential_level,
+        ).values_list("user_id", flat=True)
+
+        users_earning_credentials = (
+            CompletedWork.objects.filter(
+                opportunity_access__opportunity__id__in=opportunities,
+                status=CompletedWorkStatus.approved,
+            )
+            .exclude(opportunity_access__user_id__in=Subquery(user_credentials_to_exclude))
+            .values("opportunity_access__user_id", "opportunity_access__opportunity_id")
+            .annotate(
+                deliveries_count=Count("opportunity_access__user_id"),
+                opportunity_id=F("opportunity_access__opportunity_id"),
+                delivery_type_id=F("opportunity_access__opportunity__delivery_type_id"),
+                cred_user_id=F("opportunity_access__user_id"),
+            )
+            .filter(deliveries_count__gte=level_int)
+            .values("cred_user_id", "opportunity_id", "delivery_type_id")
+            .distinct()
+        )
+
+        return cls._parse_to_user_credential_models(
+            credentials_users=users_earning_credentials,
+            credential_type=UserCredential.CredentialType.DELIVERY,
+            level=credential_level,
+        )
+
+    @classmethod
+    def get_learning_user_credentials(cls, opportunities, credential_level):
+        user_credentials_to_exclude = UserCredential.objects.filter(
+            opportunity_id=OuterRef("opportunity_id"),
+            credential_type=UserCredential.CredentialType.LEARN,
+            level=credential_level,
+        ).values_list("user_id", flat=True)
+
+        users_earning_credentials = (
+            Assessment.objects.filter(
+                opportunity_id__in=opportunities,
+                passed=True,
+                opportunity_access__isnull=False,
+            )
+            .exclude(opportunity_access__user_id__in=Subquery(user_credentials_to_exclude))
+            .annotate(
+                delivery_type_id=F("opportunity__delivery_type_id"),
+                cred_user_id=F("opportunity_access__user_id"),
+            )
+            .values("cred_user_id", "opportunity_id", "delivery_type_id")
+            .distinct()
+        )
+        return cls._parse_to_user_credential_models(
+            credentials_users=users_earning_credentials,
+            credential_type=UserCredential.CredentialType.LEARN,
+            level=credential_level,
+        )
+
+    @classmethod
+    def _parse_to_user_credential_models(cls, credentials_users, credential_type, level):
+        return [
+            UserCredential(
+                **{
+                    "user_id": cred_user["cred_user_id"],
+                    "credential_type": credential_type,
+                    "level": level,
+                    "opportunity_id": cred_user["opportunity_id"],
+                    "delivery_type_id": cred_user.get("delivery_type_id", None),
+                }
+            )
```

Comment:
Just a query, is the dictionary unpacking needed here for some reason?

Reply:
Nope, updated here: https://github.com/dimagi/commcare-connect/pull/795/commits/b6529ebc4ad1028503718be4867b7587ce757353